DOCSP-41508: TLS (#37)
mongodb/docs-kotlin-sync: source/connect/tls.txt

.. _kotlin-sync-tls:

==========================
Enable TLS on a Connection
==========================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, security, authentication

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the 
:wikipedia:`TLS <w/index.php?title=Transport_Layer_Security&oldid=1239598620>`
security protocol when connecting to MongoDB by using the {+driver-short+}.

.. note:: Debugging TLS

   If you experience trouble setting up your TLS connection, you can
   use the ``-Djavax.net.debug=all`` system property to view helpful
   log statements. See `Debugging SSL/TLS connections
   <https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ReadDebug.html>`__

>    in the Java language documentation for more information.

Reviewer: S; i think this note is coming too early in the page. Could move to the end of the **Enable TLS** section?

.. _kotlin-sync-tls-enable:

Enable TLS
----------

You can enable TLS on a connection to your MongoDB instance
in the following ways:

- Setting the ``tls`` parameter in your connection string
- Using the ``enabled()`` method from the ``SslSettings.Builder`` class when creating a
  ``MongoClientSettings`` instance

.. note:: DNS Seedlist Protocol Enables TLS

   If you connect by using the DNS seedlist protocol, indicated by the
   ``mongodb+srv`` prefix in your connection string, the driver
   automatically enables TLS.

   To learn more about connection behavior when you use a DNS seedlist,
   see the :manual:`SRV Connection Format </reference/connection-string/#srv-connection-format>`
   section of the Connection Strings guide in the Server manual.

.. tabs::

   .. tab:: Connection String
      :tabid: connectionstring

>       To enable TLS on a connection by using a connection string, set the connection string
>       parameter ``tls`` to ``true`` in the connection string passed to
>       ``MongoClient.create()``, as shown in the following code:

Reviewer: S: reduce the number of times "connection string" is mentioned:
```suggestion
      To enable TLS on a connection by using a connection string, set the
      ``tls`` option to ``true`` in the options parameter and pass the string to
      ``MongoClient.create()``, as shown in the following code:
```

      To enable TLS within a ``MongoClientSettings`` instance, use the
      ``applyToSslSettings()`` builder method. Set the ``enabled`` property to ``true``
      in the ``SslSettings.Builder`` block, as shown in the following code:

      .. literalinclude:: /includes/connect/tls.kt
         :start-after: start-tls-mongo-client-settings
         :end-before: end-tls-mongo-client-settings
         :language: kotlin
         :copyable:
         :dedent:

.. _tls_configure-certificates:

Configure Certificates
----------------------

{+language+} applications that initiate TLS requests require access to
cryptographic certificates that prove the application's identity and verify
other applications with which the {+language+} application interacts. You can configure
access to these certificates in your application in the following ways:

- Using a JVM trust store and JVM key store
- Using a client-specific trust store and key store

.. _kotlin-sync-tls-configure-jvm-truststore:

Configure the JVM Trust Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note::

   By default, the JRE includes many commonly used public certificates
   from signing authorities such as `Let's Encrypt

>    <https://letsencrypt.org/>`__. As a result, you can connect to a
>    :atlas:`MongoDB Atlas </>` instance, or any other
>    server whose certificate is signed by an authority in the JRE's default
>    certificate store, with TLS enabled without configuring the trust store.

Reviewer: ```suggestion
   <https://letsencrypt.org/>`__. As a result, you can enable TLS when connecting to a
   :atlas:`MongoDB Atlas </>` instance, or any other
   server whose certificate is signed by an authority in the JRE's default
   certificate store, without configuring the trust store.
```

- ``javax.net.ssl.trustStore``: Path to a trust store containing the client's TLS
  certificates

- ``javax.net.ssl.trustStorePassword``: Password to access the trust
  store defined in ``javax.net.ssl.trustStore``

These properties ensure that your application can
validate the TLS certificate presented by a connected MongoDB instance.

You can create a trust store by using the `keytool <https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
command line tool from the JDK as shown in the following terminal command:

.. code-block:: console

   keytool -importcert -trustcacerts -file <path to certificate authority file>
            -keystore <path to trust store> -storepass <password>

Configure the JVM Key Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note::

   By default, MongoDB instances do not perform client certificate
   validation. You must configure the key store if you configured your MongoDB
   instance to validate client certificates.

The JVM key store saves certificates that securely identify your {+language+}
application to other applications. By using these certificates, other
applications can prove that the connection to your application is

> genuine and secure from tampering by third parties.

Reviewer: S: I think this is a duplicate paragraph from the prev section. Consider removing

An application that initiates TLS requests must set the following JVM system
properties to ensure that the client presents a TLS certificate to
the MongoDB server:

- ``javax.net.ssl.keyStore``: Path to a key store containing the client's
  TLS/SSL certificates

- ``javax.net.ssl.keyStorePassword``: Password to access the key store
  defined in ``javax.net.ssl.keyStore``

You can create a key store by using the `keytool
<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
or `openssl <https://www.openssl.org/docs/manmaster/man1/openssl.html>`__
command line tool.

To learn more about configuring a {+language+} application to use TLS,
see the `JSSE Reference Guide <https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html>`__
in the Java language documentation.

.. _tls-disable-hostname-verification:

Configure a Client-Specific Trust Store and Key Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can configure a client-specific trust store and key store by using the
``init()`` method of the ``SSLContext`` class.

> Find an example showing how to configure a client to use an ``SSLContext``
> instance in the :ref:`Customize TLS Configuration with an SSLContext section of this guide <kotlin-sync-tls-custom-sslContext>`.

Reviewer: ```suggestion
Find an example showing how to configure a client to use an ``SSLContext``
instance in the :ref:`kotlin-sync-tls-custom-sslContext` section of this guide.
```

Disable Hostname Verification
-----------------------------

By default, the driver ensures that the hostname included in the server's
TLS certificates matches the hostnames provided when constructing
a ``MongoClient``. To disable hostname verification for your
application, set the ``invalidHostNameAllowed`` property of the builder to ``true`` in the
``applytoSslSettings()`` builder block:

.. literalinclude:: /includes/connect/tls.kt
   :start-after: start-disable-hostname-verification
   :end-before: end-disable-hostname-verification
   :language: kotlin
   :copyable:
   :dedent:

.. warning::

   Disabling hostname verification makes your application insecure and potentially
   vulnerable to expired certificates and foreign processes posing as valid client
   instances.

.. _kotlin-sync-tls-restrict-tls-1.2:

Restrict Connections to TLS 1.2 Only
------------------------------------

> To restrict your application to use only the TLS 1.2 protocol, set the
> ``jdk.tls.client.protocols`` system property to "TLSv1.2".

Reviewer: ```suggestion
To restrict your application to use only the TLS 1.2 protocol, set the
``jdk.tls.client.protocols`` system property to ``"TLSv1.2"``.
```

either the `ConnectionString <{+core-api+}/com/mongodb/ConnectionString.html>`__
or `MongoClientSettings <{+core-api+}/com/mongodb/MongoClientSettings.html>`__.

.. note:: Debugging TLS/SSL

   If you experience trouble setting up your TLS/SSL connection, you can
   use the ``-Djavax.net.debug=all`` system property to view more
   log statements. See `the Oracle guide to debugging TLS/SSL connections
   <https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ReadDebug.html>`__
   for more information.

.. _tls-enable:

Enable TLS/SSL
--------------

You can enable TLS/SSL for the connection to your MongoDB instance
in two different ways: through a parameter in your connection string, or
using a method in the ``MongoClientSettings.Builder`` class.

.. note:: DNS Seedlist Protocol Enables TLS

   If you connect by using the DNS seedlist protocol, indicated by the
   ``mongodb+srv`` prefix in your connection string, the driver
   automatically enables TLS/SSL. To disable it, set the ``tls``
   parameter value to ``false`` in your connection string, or set the
   ``enabled`` property to ``false`` in the ``SslSettings.Builder``

>    block when creating a ``MongoClientSettings`` instance.

Author: Can remove the implementation details!

   To learn more about connection behavior when you use a DNS seedlist,
   see the :manual:`SRV Connection Format </reference/connection-string/#std-label-connections-dns-seedlist>`
   section in the Server manual.

.. tabs::

   .. tab:: ConnectionString
      :tabid: connectionstring

>       To enable TLS/SSL on a connection with a `ConnectionString

Author: Will update

      ``MongoClientSettings.Builder`` class, call the
      `applyToSslSettings() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.Builder.html#applyToSslSettings(com.mongodb.Block)>`__
      method. Set the ``enabled`` property to ``true`` in the ``SslSettings.Builder``
      block to enable TLS/SSL:

      .. literalinclude:: /includes/connect/tls.kt
         :start-after: start-tls-mongo-client-settings
         :end-before: end-tls-mongo-client-settings
         :language: kotlin
         :copyable:
         :dedent:

.. _tls_configure-certificates:

Configure Certificates
----------------------

{+language+} applications that initiate TLS/SSL requests require access to
cryptographic certificates that prove identity for the application
itself and other applications with which the application
interacts. You can configure access to these certificates in your application with
the following mechanisms:

- The JVM trust store and JVM key store
- A client-specific trust store and key store

.. note::

   The following sections are based on the documentation for Oracle JDK,
   so some parts may be inapplicable to your JDK or to the custom TLS/SSL

>    implementation you use.

Author: I think we're better served just removing this.

   <https://letsencrypt.org/>`__. As a result, you can connect to
   instances of :atlas:`MongoDB Atlas </?jmp=docs_driver_kotlin>` (or any other
   server whose certificate is signed by an authority in the JRE's default
   certificate store) with TLS/SSL without configuring the trust store.

The JVM trust store saves certificates that securely identify other
applications with which your {+language+} application interacts. Using these
certificates, your application can prove that the connection to another
application is genuine and secure from tampering by third parties.

If your MongoDB instance uses a certificate that is signed by an
authority that is not present in the JRE's default certificate store,
your application must configure two system properties to initiate
SSL/TLS requests. These properties ensure that your application can
validate the TLS/SSL certificate presented by a connected MongoDB instance.

- ``javax.net.ssl.trustStore``: the path to a trust store containing the
  certificate of the signing authority

- ``javax.net.ssl.trustStorePassword``: the password to access the trust
  store defined in ``javax.net.ssl.trustStore``

> You can create a trust store with the `keytool
> <https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
> command line tool provided as part of the JDK:

Reviewer: ```suggestion
You can create a trust store by using the `keytool
<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
command line tool from the JDK as shown in the following terminal command:
```

.. code-block:: console

   keytool -importcert -trustcacerts -file <path to certificate authority file>
            -keystore <path to trust store> -storepass <password>

Configure the JVM Key Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note::

   By default, MongoDB instances do not perform client certificate
   validation. You must configure the key store if you configured your MongoDB
   instance to validate client certificates.

The JVM key store saves certificates that securely identify your {+language+}

> application to other applications. Using these certificates, other

Reviewer: ```suggestion
application to other applications. By using these certificates, other
```

applications can prove that the connection to your application is
genuine and secure from tampering by third parties.

> An application that initiates TLS/SSL requests needs to set two JVM system

Reviewer: ```suggestion
An application that initiates TLS/SSL requests must set the following JVM system
```

properties to ensure that the client presents a TLS/SSL certificate to

> the MongoDB server:

Reviewer: S: use source constant

Author: Q: referring to the "MongoDB server"? The source constant has it formatted as a proper noun with leading capitalization (`MongoDB Server`), is that okay?

Reviewer: lgtm

> - ``javax.net.ssl.keyStore``: the path to a key store containing the client's
>   TLS/SSL certificates
> 
> - ``javax.net.ssl.keyStorePassword``: the password to access the key store
>   defined in ``javax.net.ssl.keyStore``

Reviewer: ```suggestion
- ``javax.net.ssl.keyStore``: Path to a key store containing the client's
  TLS certificates
- ``javax.net.ssl.keyStorePassword``: Password to access the key store
  defined in ``javax.net.ssl.keyStore``
```

> You can create a key store with the `keytool
> <https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
> or `openssl <https://www.openssl.org/docs/manmaster/man1/openssl.html>`__
> command line tool.

Reviewer: ```suggestion
You can create a key store by using the `keytool
<https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html>`__
or `openssl <https://www.openssl.org/docs/manmaster/man1/openssl.html>`__
command line tool.
```

> For more information on configuring a {+language+} application to use TLS/SSL,
> please see the `JSSE Reference Guide
> <https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html>`__.

Reviewer: ```suggestion
To learn more about configuring a {+language+} application to use TLS,
see the `JSSE Reference Guide
<https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html>`__
in the Java language documentation.
```

.. _tls-disable-hostname-verification:

Configure a Client-Specific Trust Store and Key Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can configure a client-specific trust store and key store using the
> ``init()`` method of the ``SSLContext`` class.

Reviewer: ```suggestion
You can configure a client-specific trust store and key store by using the
``init()`` method of the ``SSLContext`` class.
```

> You can find an example showing how to configure a client with an ``SSLContext``
> instance in the
> :ref:`Customize TLS/SSL Configuration with an SSLContext section of this guide <tls-custom-sslContext>`.

Reviewer: ```suggestion
Find an example showing how to configure a client to use an ``SSLContext``
instance in the :ref:`Customize TLS/SSL Configuration with an SSLContext section of this guide <tls-custom-sslContext>`.
```

For more information on the ``SSLContext`` class, see the API

> documentation for `SSL Context <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/javax/net/ssl/SSLContext.html>`__.

Reviewer: S: should this link be moved to where there is a code example that uses it?

Author: Will do

Disable Hostname Verification
-----------------------------

By default, the driver ensures that the hostname included in the server's
TLS/SSL certificates matches the hostnames provided when constructing

> a ``MongoClient``. To disable hostname verification for your
> application, you can explicitly disable this by setting the
> ``invalidHostNameAllowed`` property of the builder to ``true`` in the
> ``applytoSslSettings()`` builder lambda:

Reviewer: ```suggestion
a ``MongoClient``. To disable hostname verification for your
application, set the
``invalidHostNameAllowed`` property of the builder to ``true`` in the
``applytoSslSettings()`` builder block:
```

.. literalinclude:: /includes/connect/tls.kt
   :start-after: start-disable-hostname-verification
   :end-before: end-disable-hostname-verification
   :language: kotlin
   :copyable:
   :dedent:

.. warning::

   Disabling hostname verification can make your configuration
   `insecure <https://tlseminar.github.io/docs/mostdangerous.pdf>`__.
   Disable hostname verification only for testing purposes or

>    when there is no other alternative.

Reviewer: Specifying this option in a production environment makes your application insecure and potentially vulnerable to expired certificates and foreign processes posing as valid client instances.

Reviewer: from the GO copy

Author: Will modify the warning.

> .. _tls-restrict-tls-1.2:

Reviewer: S: see comment about using language in all anchors

Restrict Connections to TLS 1.2 Only
------------------------------------

To restrict your application to use only the TLS 1.2 protocol, set the
``jdk.tls.client.protocols`` system property to "TLSv1.2".

.. note::

   Java Runtime Environments (JREs) before Java 8 only enabled
   the TLS 1.2 protocol in update releases. If your JRE has not enabled
   the TLS 1.2 protocol, upgrade to a later release to connect by using
   TLS 1.2.

.. _tls-custom-sslContext:

Customize TLS/SSL Configuration through the Java SE SSLContext
--------------------------------------------------------------

> If your TLS/SSL configuration requires customization, you can
> set the ``sslContext`` property of your ``MongoClient`` by
> passing an `SSLContext
> <https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html>`__
> object to the builder in the ``applyToSslSettings()`` lambda:

Reviewer: ```suggestion
If your TLS/SSL configuration requires customization, you can
set the ``sslContext`` property of your ``MongoClient`` by
passing an `SSLContext
<https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html>`__
object to the ``context()`` method builder in the ``applyToSslSettings()`` block:
```

.. literalinclude:: /includes/connect/tls.kt
   :start-after: start-ssl-context
   :end-before: end-ssl-context
   :language: kotlin
   :copyable:
   :dedent:

Online Certificate Status Protocol (OCSP)
-----------------------------------------

OCSP is a standard used to check whether X.509 certificates have been
revoked. A certificate authority can add an X.509 certificate to the
Certificate Revocation List (CRL) before the expiry time to invalidate
the certificate. When a client sends an X.509 certificate during the TLS
handshake, the CA's revocation server checks the CRL and returns a status

> of "good", "revoked", or "unknown".

Reviewer: ```suggestion
of ``good``, ``revoked``, or ``unknown``.
```

The driver supports the following variations of OCSP:

- **Client-Driven OCSP**

> - **OCSP Stapling**

Reviewer: S: unbold
